README.md

```diff
@@ -177,44 +177,64 @@ Example of a partial CR:
 
 ### Mounting and injecting binding information
 
-This specification allows for data to be mounted using volumes or injected using environment variables.  The best practice is to mount any sensitive information, such as passwords, since that will avoid accidentally exposure via environment dumps and subprocesses.  Also, binding binary data (e.g. .p12 certificate for Kafka) as an environment variable might cause a pod to fail to start (stuck on `CrashLoopBackOff`), so it advisable for backing services with such binding data to mark it with `bindAs: volume`.
+#### Mounting data
+Implementations of this specification **MUST** mount the binding data into the consuming application container in the following location:
+
+```
+<mountPathPrefix>/<service-id>/secret/<persisted_secret>
+```
+
+Where:
+* `<mountPathPrefix>` defaults to `/platform/bindings` if not specified in the `ServiceBinding` CR via the `mountPathPrefix` element.  The environment variable `SERVICE_BINDINGS_ROOT` **MUST** be set to the value of `<mountPathPrefix>`, so that applications can always find this directory, and it is immutable.  This means that if another `ServiceBinding` CR wants to project itself into the same container, it **MUST** reuse the current `SERVICE_BINDINGS_ROOT` value.  
+* `<service-id>` is the `id` field of the corresponding `service` entry in the `ServiceBinding` CR.  If the `id` field is not present, the `name` field is used instead.  The `<service-id>` path **MUST** be unique between the services bound to a particular application.
+* `<persisted_secret>` represents a set of files where the filename is a Secret key and the file contents is the corresponding value of that key.
+
+Example:  `/platform/bindings/mongo-db/secret/MONGODB_HOST`
+
+#### Custom bindings
```

Comment:
I don't believe that this needs to be reserved for custom usage.  If you think about the projection of a binding into the container as being loosely coupled to what has provided the metadata/secret for it, then all of the binding ids are equivalent and any of them could be "custom"; the application in the container would never need to know.  A binding projection could be:

* a perfect copy of a binding outside the container
* a copy of a binding outside the container with the name changed
* a completely custom configuration with a `ConfigMap` and `Secret` defined on a `Deployment` (just handwritten into a Spec)
* a custom configuration that makes reference to other `ConfigMap`s and `Secret`s some hand written, some from bindings outside the container.

I think this section should be removed as it's a special case of something that doesn't need to be special-cased.

Reply:
yeah, this is definitely an area where we can give more thought.    

Historically, our flow was that all of the service bindings that the app "sees" are:
(1) the bindings we fetched from the services defined in the `ServiceBindingRequest` CR 
(2)the customized `dataMappings`, which included renamed items as well as cross-service composed bindings.

They were always persisted as a single list of bindings.  

Now that we have switched to a model where the bindings are persisted in a path that includes the `service-id`, we have this issue to solve where the `dataMappings` elements from our `ServiceBindingRequest` CR can be renames of multiple services.  

One suggestion that @yharish991 had is to move the `dataMappings` underneath each of the `services` entry - which would solve the problem on "where" we persist that (i.e. under the corresponding `service-id`.   But we still have issues with composed bindings (aggregating from multiple services).  Where would that be persisted?

Reply:
I think there's a conflation of the actors involved here.  From the view of the projection as represented on the PodSpec, there are just "some volumes":

```yaml
spec:
  containers:
  - name: app
    image: nginx
    env:
    - name: SERVICE_BINDING_ROOT
      value: /platform/bindings
    volumeMounts:
    - name: my-nosql-db-metadata
       mountPath:  /platform/bindings/my-nosql-db/metadata
       readOnly: true
    - name: my-nosql-db-secret
       mountPath:  /platform/bindings/my-nosql-db/secret
       readOnly: true
  volumes:
  - name: my-nosql-db-metadata
     configMap:
       name: my-mongodb-metadata
  - name: my-nosql-db-secret
     secret:
       secretRef: my-mongodb-secret
```

How those volumes got reconciled to there, what their names are and how those were calculated, don't really matter in the context of this part of the spec.  What makes this so nice is that *another* reconciler can come along and do something fun like adding a binding that is synthesized from other bindings.

```yaml
spec:
  containers:
  - name: app
    image: nginx
    env:
    - name: SERVICE_BINDING_ROOT
      value: /platform/bindings
    volumeMounts:
    - name: my-nosql-db-metadata
       mountPath:  /platform/bindings/my-nosql-db/metadata
       readOnly: true
    - name: my-nosql-db-secret
       mountPath:  /platform/bindings/my-nosql-db/secret
       readOnly: true

    - name: my-rdbms-metadata
       mountPath:  /platform/bindings/my-rdbms/metadata
       readOnly: true
    - name: my-rdbms-secret
       mountPath:  /platform/bindings/my-rdbms/secret
       readOnly: true
  volumes:
  - name: my-nosql-db-metadata
     configMap:
       name: my-mongodb-metadata
  - name: my-nosql-db-secret
     secret:
       secretRef: my-mongodb-secret

  - name: my-rdbms-metadata
     configMap:
       name: my-custom-metadata
  - name: my-rdbms-secret
     secret:
       secretRef: my-custom-secret

---
apiVersion: v1
kind: ConfigMap
metadata:
  name: my-custom-metadata
data:
  kind: mysql
---
apiVersion: v1
kind: Secret
metadata:
  name: my-custom-secret
type: Opaque
stringData:  # This data is synthesized by some other reconciler
  jdbcUrl: jdbc:mysql://${mysql.secret.host}:${mysql.secret.port}/${mysql.secret.database}?reconnect=true
  password: ${mysql.secret.password}
  username: ${mysql.secret.username}
```

Key here though, is that this isn't a spec at all.  It's just another pair of volumes, mounted to a predictable location, by some other reconciler (essentially a distributed model of binding generation).

Finally, there's nothing to stop a user from *just creating a hand-rolled-binding*:

```yaml
spec:
  containers:
  - name: app
    image: nginx
    env:
    - name: SERVICE_BINDING_ROOT
      value: /platform/bindings
    volumeMounts:
    - name: my-nosql-db-metadata
       mountPath:  /platform/bindings/my-nosql-db/metadata
       readOnly: true
    - name: my-nosql-db-secret
       mountPath:  /platform/bindings/my-nosql-db/secret
       readOnly: true

    - name: my-rdbms-metadata
       mountPath:  /platform/bindings/my-rdbms/metadata
       readOnly: true
    - name: my-rdbms-secret
       mountPath:  /platform/bindings/my-rdbms/secret
       readOnly: true

    - name: my-new-relic-metadata
       mountPath:  /platform/bindings/my-new-relic/metadata
       readOnly: true
    - name: my-new-relic-secret
       mountPath:  /platform/bindings/my-new-relic/secret
       readOnly: true
  volumes:
  - name: my-nosql-db-metadata
     configMap:
       name: my-mongodb-metadata
  - name: my-nosql-db-secret
     secret:
       secretRef: my-mongodb-secret

  - name: my-rdbms-metadata
     configMap:
       name: my-custom-metadata
  - name: my-rdbms-secret
     secret:
       secretRef: my-custom-secret

  - name: my-new-relic-metadata
     configMap:
       name: my-hand-rolled-metadata
  - name: my-new-relic-secret
     secret:
       secretRef: my-hand-rolled-secret
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: my-custom-metadata
data:
  kind: mysql
---
apiVersion: v1
kind: Secret
metadata:
  name: my-custom-secret
type: Opaque
stringData:  # This data is synthesized by some other reconciler
  jdbcUrl: jdbc:mysql://${mysql.secret.host}:${mysql.secret.port}/${mysql.secret.database}?reconnect=true
  password: ${mysql.secret.password}
  username: ${mysql.secret.username}

---
apiVersion: v1
kind: ConfigMap
metadata:
  name: my-hand-rolled-metadata
data:
  kind: new-relic
---
apiVersion: v1
kind: Secret
metadata:
  name: my-hand-rolled-secret
type: Opaque
stringData:  # This data is hand entered by a user
  licenseKey: <NEW-RELIC-LICENSE-KEY>
```

Reply:
awesome samples @nebhale!  thanks.  We should add some of these to the spec.  :)